### ISSUE TITLE
Mypy type hint error

### ISSUE THREAD

**Issue by Kontrolix** [alecthomas/voluptuous#491]
With the new release 0.14.0, we experienced a lot of mypy errors due to a bad type hint such as:

```
error: Argument 1 to "Coerce" has incompatible type "Callable[[Any, int, bool], Any]"; expected "type"  [arg-type]
error: Value of type "Schema" is not indexable  [index]
error: Unsupported target for indexed assignment ("Schema")  [index]
error: Returning Any from function declared to return "Schema"  [no-any-return]
```

I tried to fix the Coerce one [here](https://github.com/alecthomas/voluptuous/pull/488) but I am facing some compatibility issues in the CI with python 3.7, 3.8, 3.9.

If we want to support mypy and type hint, do we really want to continue to support these old versions?

**Comment by antoni-szych-rtbhouse** [comment]
side note:
- 3.6 support was removed in 0.14.0.
- [3.7 reached end of life 2023-06-27](https://peps.python.org/pep-0537/#lifespan)
- 3.8 will [reach EOL in October 2024](https://peps.python.org/pep-0569/#lifespan)

**Comment by alecthomas** [comment]
I'm fine with dropping support for EOL Python versions.

**Comment by spacegaier** [comment]
@Kontrolix With the recent fixes in 0.15.x, are you still facing issues or can this issue be closed?

**Comment by Kontrolix** [comment]
Sorry, but I'm no longer working for the company where I met those errors. So I can't tell you.

I gess we can close this issue.
